src/ray/object_manager/object_manager.cc

```diff
@@ -164,287 +317,140 @@ ray::Status ObjectManager::Wait(const std::vector<ObjectID> &object_ids,
-  boost::asio::read(conn->GetSocket(), header, ec);
-  // read data
-  std::vector<uint8_t> message;
-  message.resize(length);
-  boost::asio::read(conn->GetSocket(), boost::asio::buffer(message), ec);
-  // Serialize
-  auto info = flatbuffers::GetRoot<ClientConnectionInfo>(message.data());
+void ObjectManager::DisconnectClient(std::shared_ptr<ReceiverConnection> &conn,
+                                     const uint8_t *message) {
+  auto info = flatbuffers::GetRoot<DisconnectClientMessage>(message);
   ClientID client_id = ObjectID::from_binary(info->client_id()->str());
   bool is_transfer = info->is_transfer();
-  // TODO: trash connection if either fails.
   if (is_transfer) {
-    transfer_receive_connections_[client_id] = conn;
-    Status status = WaitPushReceive(conn);
-    return status;
+    connection_pool_.RemoveReceiver(ConnectionPool::TRANSFER, client_id, conn);
   } else {
-    message_receive_connections_[client_id] = conn;
-    Status status = WaitMessage(conn);
-    return status;
+    connection_pool_.RemoveReceiver(ConnectionPool::MESSAGE, client_id, conn);
   }
 };
 
-ray::Status ObjectManager::WaitPushReceive(TCPClientConnection::pointer conn) {
-  boost::asio::async_read(
-      conn->GetSocket(),
-      boost::asio::buffer(&conn->message_length_, sizeof(conn->message_length_)),
-      boost::bind(&ObjectManager::HandlePushReceive, this, conn,
-                  boost::asio::placeholders::error));
+void ObjectManager::ReceivePullRequest(std::shared_ptr<ReceiverConnection> &conn,
+                                       const uint8_t *message) {
+  // Serialize.
+  auto pr = flatbuffers::GetRoot<PullRequestMessage>(message);
+  ObjectID object_id = ObjectID::from_binary(pr->object_id()->str());
+  ClientID client_id = ClientID::from_binary(pr->client_id()->str());
+  // Push object to requesting client.
+  ray::Status push_status = Push(object_id, client_id);
+};
+
+ray::Status ObjectManager::WaitPushReceive(std::shared_ptr<ReceiverConnection> conn) {
+  //  RAY_LOG(INFO) << "WaitPushReceive";
+  asio::async_read(conn->GetSocket(), asio::buffer(&read_length_, sizeof(read_length_)),
+                   boost::bind(&ObjectManager::HandlePushReceive, this, conn,
+                               asio::placeholders::error));
   return ray::Status::OK();
-}
+};
 
-void ObjectManager::HandlePushReceive(TCPClientConnection::pointer conn,
-                                      BoostEC length_ec) {
+void ObjectManager::HandlePushReceive(std::shared_ptr<ReceiverConnection> conn,
+                                      const boost::system::error_code &length_ec) {
+  //  RAY_LOG(INFO) << "HandlePushReceive";
   std::vector<uint8_t> message;
-  message.resize(conn->message_length_);
+  message.resize(read_length_);
   boost::system::error_code ec;
-  boost::asio::read(conn->GetSocket(), boost::asio::buffer(message), ec);
+  asio::read(conn->GetSocket(), asio::buffer(message), ec);
```

Comment:
We should really be using the common `ClientConnection` class for reading messages from the socket like this! Right now, it seems like all connections to an object manager will share the same `read_length_` variable. I am pretty sure this will break once we receive multiple concurrent `Push`es.

Reply:
I've corrected this and added a test to cover the failure scenario.